File: advent_of_code_clj/aoc_2020/src/aoc_2020/day05.clj
(ns aoc-2020.day05
  (:require [aoc-utils.core :as utils]))

;; low and high will always have a different parity hence the + 1
(defn- split [{:keys [low high] :as acc} c]
  (cond
    (or (= c \F) (= c \L)) (update acc
                                   :high (fn [_] (dec (/ (+ low high 1) 2))))
    (or (= c \B) (= c \R))  (update acc
                                    :low (fn [_] (/ (+ low high 1) 2)))))

#_(split {:low 0 :high 127} \F)
#_(split (split {:low 0 :high 127} \F) \B)
#_(split {:low 0 :high 127} \B)

(defn- simplify [l high]
  (loop [[hd & tl] l,
         acc {:low 0, :high high}]
    (cond
      (nil? hd) (:low acc)
      :else (recur tl (split acc hd)))))

#_(simplify '(\F \B \F \B \B \F \F) 127)

(defn- id [pass]
  (let [row (take 7 pass)
        col (drop 7 pass)]
    (+ (* 8 (simplify row 127)) (simplify col 7))))

#_(id "FBFBBFFRLR")

(defn part-1 [file]
  (utils/reduce-file file (fn [acc line] (max acc (id line))) 0))

#_(part-1 (clojure.java.io/resource "day05-example"))

(defn- find-gap [[hd & tl]]
  (loop [prev hd,
         [hd & tl] tl]
    (cond
      (nil? hd) (prn "error")
      (= (inc prev) hd) (recur hd tl)
      :else (dec hd))))

#_(find-gap '(1 2 3 4 6 7))

(defn part-2 [file]
  (as->
   (utils/reduce-file file (fn [acc line] (conj acc (id line))) #{}) set
    (sort set)
    (find-gap set)))

#_(part-2 (clojure.java.io/resource "day05"))

File: advent_of_code_ocaml/2020/src/days/day05.ml
open Mdrp_lib

let split (low, high) = function
  | 'F' | 'L' -> (low, (low + high) / 2)
  | 'B' | 'R' -> ((low + high) / 2, high)
  | _ -> assert false

let parse s =
  String.foldi
    (fun i (row, col) c ->
      if i < 7 then (split row c, col) else (row, split col c))
    ((0, 127), (0, 7))
    s

let part_1 file =
  Parse.fold_lines
    (fun acc line ->
      let (_, row), (_, col) = parse line in
      max acc ((row * 8) + col))
    0 file

exception Found of int

let part_2 file =
  let set =
    Parse.fold_lines
      (fun acc line ->
        let (_, row), (_, col) = parse line in
        Int.Set.add ((row * 8) + col) acc)
      Int.Set.empty file
  in
  try
    Int.Set.fold
      (fun e prev -> if e = prev + 1 then e else raise (Found (e - 1)))
      set
      (Int.Set.min_elt set - 1)
  with Found i -> i

let run part file = match part with 1 -> part_1 file | _ -> part_2 file
